library/src/main/java/com/datatorrent/lib/join/AbstractManagedStateInnerJoinOperator.java

```diff
@@ -0,0 +1,354 @@
+    ((FileAccessFSImpl)stream2Store.getFileAccess()).setBasePath(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + stateDir + Path.SEPARATOR + String.valueOf(context.getId()) + Path.SEPARATOR + stream2State);
+    stream1Store.getCheckpointManager().setRecoveryPath("managed_state_" + stream1State);
+    stream1Store.getCheckpointManager().setRecoveryPath("managed_state_" + stream2State);
+    stream1Store.setup(context);
+    stream2Store.setup(context);
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    stream1Store.beginWindow(windowId);
+    stream2Store.beginWindow(windowId);
+    super.beginWindow(windowId);
+  }
+
+  /**
+   * Process the waiting events
+   * @param finalize finalize Whether or not to wait for future to return
+   */
+  private void processWaitEvents(boolean finalize)
+  {
+    Iterator<Map.Entry<JoinEvent<K,T>, Future<List>>> waitIterator = waitingEvents.entrySet().iterator();
+    while (waitIterator.hasNext()) {
+      Map.Entry<JoinEvent<K,T>, Future<List>> waitingEvent = waitIterator.next();
+      Future<List> future = waitingEvent.getValue();
+      if (future.isDone() || finalize) {
+        try {
+          JoinEvent<K,T> event = waitingEvent.getKey();
+          joinStream(event.value,event.isStream1Data,future.get());
+        } catch (InterruptedException | ExecutionException e) {
+          throw new RuntimeException("end window", e);
+        }
+        waitIterator.remove();
+        if (!finalize) {
+          break;
+        }
+      }
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    processWaitEvents(true);
+    stream1Store.endWindow();
+    stream2Store.endWindow();
+    super.endWindow();
+  }
+
+  @Override
+  public void teardown()
+  {
+    stream1Store.teardown();
+    stream2Store.teardown();
+    super.teardown();
+  }
+
+  /**
+   * Return the number of buckets
+   * @return the noOfBuckets
+   */
+  public int getNoOfBuckets()
+  {
+    return noOfBuckets;
+  }
+
+  /**
+   * Set the number of buckets required for managed state
+   * @param noOfBuckets noOfBuckets
+   */
+  public void setNoOfBuckets(int noOfBuckets)
+  {
+    this.noOfBuckets = noOfBuckets;
+  }
+
+  /**
+   * Return the bucketSpanTime
+   * @return the bucketSpanTime
+   */
+  public Long getBucketSpanTime()
+  {
+    return bucketSpanTime;
+  }
+
+  /**
+   * Sets the length of the time bucket required for managed state.
+   * @param bucketSpanTime given bucketSpanTime
+   */
+  public void setBucketSpanTime(Long bucketSpanTime)
+  {
+    this.bucketSpanTime = bucketSpanTime;
+  }
+
+  public static class JoinEvent<K,T>
+  {
+    public K key;
+    public T value;
+    public boolean isStream1Data;
+
+    public JoinEvent(K key, T value, boolean isStream1Data)
+    {
+      this.key = key;
+      this.value = value;
+      this.isStream1Data = isStream1Data;
+    }
+  }
+
+  public class ManagedSpillableComplexComponent implements SpillableComplexComponent
```

Comment:
Do you need this complete class to be implemented? 
Can you not do:
new ManagedTimeStateMultiValue<K,V>(store, isPrimaryKey);
new ManagedTimeStateMultiValue<K,V>();

Wherever required?